File: app/src/main/java/in/iot/lab/dmscanner/MainActivity2.java
package in.iot.lab.dmscanner;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import org.libdmtx.DMTXImage;
import org.libdmtx.DMTXTag;

import java.io.FileNotFoundException;
import java.io.InputStream;
import java.nio.IntBuffer;

import in.iot.lab.dmscanner.databinding.ActivityMainBinding;

public class MainActivity2 extends AppCompatActivity {

    private static final int REQ_CODE_PICK_IMAGE = 0;
    private static final int REQ_CODE_CAMERA = 1;
    private ActivityMainBinding mainBinding;
    private static final int RC_PERMISSION = 10;
    private CodeScanner mCodeScanner;
    private boolean mPermissionGranted;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mainBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mainBinding.getRoot());
        mCodeScanner = new CodeScanner(this, mainBinding.scannerView);
        mCodeScanner.setDecodeCallback(result -> runOnUiThread(() -> {
//            ScanResultDialog dialog = new ScanResultDialog(this, result);
//            dialog.setOnDismissListener(d -> mCodeScanner.startPreview());
//            dialog.show();
//            Toast.makeText(this," "+result.getText(),Toast.LENGTH_LONG).show();
            mainBinding.textView.setText(result.getText());
//            mCodeScanner.startPreview();
        }));
        mCodeScanner.setErrorCallback(error -> runOnUiThread(
                () -> Toast.makeText(this, "Error in scanning!!", Toast.LENGTH_LONG).show()));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                mPermissionGranted = false;
                requestPermissions(new String[]{Manifest.permission.CAMERA}, RC_PERMISSION);
            } else {
                mPermissionGranted = true;
            }
        } else {
            mPermissionGranted = true;
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        mainBinding.textView.setOnClickListener(view -> {
            showChooser();
        });
        if (mPermissionGranted) {
            mCodeScanner.startPreview();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        if (requestCode == RC_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                mPermissionGranted = true;
                mCodeScanner.startPreview();
            } else {
                mPermissionGranted = false;
            }
        }
    }

    @Override
    protected void onPause() {
        mCodeScanner.releaseResources();
        super.onPause();
    }

    public void showChooser() {
        Intent photoPickerIntent = new Intent(Intent.ACTION_GET_CONTENT);
        photoPickerIntent.setType("*/*");
        photoPickerIntent.addCategory(Intent.CATEGORY_OPENABLE);
        startActivityForResult(
                Intent.createChooser(photoPickerIntent, "Select a File to Upload"),
                REQ_CODE_PICK_IMAGE);
    }

    private void decode(Bitmap img) {
        System.out.println("DECODING");
        int width = img.getWidth();
        int height = img.getHeight();
        System.out.println(img.getWidth());
        System.out.println(img.getHeight());
        Bitmap scaledBitmap = Bitmap.createScaledBitmap(img, (int) (img.getWidth() * 0.5), (int) (img.getHeight() * 0.5), true);
        int midx = scaledBitmap.getWidth() / 2;
        int midy = scaledBitmap.getHeight() / 2;
//    	Bitmap cropImg = Bitmap.createBitmap(scaledBitmap, midx - width / 2, midy - height/2, width, height);
        int size = scaledBitmap.getHeight() * scaledBitmap.getWidth();
        IntBuffer buff = IntBuffer.allocate(size);
        scaledBitmap.copyPixelsToBuffer(buff);
        DMTXImage dmtxImage = new DMTXImage(scaledBitmap.getWidth(), scaledBitmap.getHeight(), buff.array());
        DMTXTag[] tags = dmtxImage.getTags(5, 10000);
        System.out.println(tags.length);
        StringBuilder sb = new StringBuilder();
        sb.append(tags.length + " tags found\n");
        for (DMTXTag tag : tags) {
            sb.append(tag.id + "\n");
            System.out.println(tag.id);
        }
        String text = sb.toString();
        TextView textView = findViewById(R.id.textView);
        textView.setText(text);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
        super.onActivityResult(requestCode, resultCode, intent);
        if (requestCode == REQ_CODE_PICK_IMAGE) {
            if (resultCode == RESULT_OK) {
                Uri selectedImage = intent.getData();
                try {
                    InputStream inputStream = getContentResolver().openInputStream(selectedImage);
                    Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
                    decode(bitmap);
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                }

            }
        } else if (requestCode == REQ_CODE_CAMERA) {
            Bitmap img = (Bitmap) intent.getExtras().get("data");
            decode(img);
        }
    }
}

File: app/src/main/java/in/iot/lab/dmscanner/ui/MainActivity2.kt
package `in`.iot.lab.dmscanner.ui

import `in`.iot.lab.dmscanner.R
import `in`.iot.lab.dmscanner.databinding.ActivityMainBinding
import `in`.iot.lab.dmscanner.databinding.QrScannedBinding
import `in`.iot.lab.dmscanner.scanner.CodeScanner
import `in`.iot.lab.dmscanner.scanner.ErrorCallback
import `in`.iot.lab.dmscanner.scanner.Utils
import `in`.iot.lab.dmscanner.model.firebaseDataInput
import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.bottomsheets.BottomSheet
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.google.zxing.Result
import org.opencv.android.OpenCVLoader
import java.io.FileNotFoundException

class MainActivity2 : AppCompatActivity() {
    private var fireBaseInput=firebaseDataInput()
    private var mainBinding: ActivityMainBinding? = null
    private lateinit var mCodeScanner: CodeScanner
    private var mPermissionGranted = false
    override fun onCreate(savedInstanceState: Bundle?) {

        OpenCVLoader.initDebug()
        super.onCreate(savedInstanceState)
        mainBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(mainBinding!!.root)
        mCodeScanner = CodeScanner(this, mainBinding!!.scannerView)
        mCodeScanner.setDecodeCallback { result: Result ->
            fireBaseInput.putInDataBase(result.text.replace("1 tags found","").trim())
            runOnUiThread {
                displayQR(result.text)
            }

        }

        mCodeScanner.setErrorCallback(ErrorCallback { error: Exception? ->
            runOnUiThread {
                Toast.makeText(
                    this,
                    "Error in scanning!!",
                    Toast.LENGTH_LONG
                ).show()
            }
        })
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                mPermissionGranted = false
                requestPermissions(arrayOf(Manifest.permission.CAMERA), RC_PERMISSION)
            } else {
                mPermissionGranted = true
            }
        } else {
            mPermissionGranted = true
        }
    }

    override fun onResume() {
        super.onResume()
        mainBinding!!.textView.setOnClickListener { view: View? -> showChooser() }
        if (mPermissionGranted) {
            mCodeScanner.startPreview()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>,
        grantResults: IntArray
    ) {
        if (requestCode == RC_PERMISSION) {
            if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                mPermissionGranted = true
                mCodeScanner.startPreview()
            } else {
                mPermissionGranted = false
            }
        }
    }

    override fun onPause() {
        mCodeScanner.releaseResources()
        super.onPause()
    }

    fun showChooser() {
        val photoPickerIntent = Intent(Intent.ACTION_GET_CONTENT)
        photoPickerIntent.type = "*/*"
        photoPickerIntent.addCategory(Intent.CATEGORY_OPENABLE)
        startActivityForResult(
            Intent.createChooser(photoPickerIntent, "Select a File to Upload"),
            REQ_CODE_PICK_IMAGE
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)
        if (requestCode == REQ_CODE_PICK_IMAGE) {
            if (resultCode == RESULT_OK) {
                val selectedImage = intent!!.data
                try {
                    val inputStream = contentResolver.openInputStream(
                        selectedImage!!
                    )
                    val bitmap = BitmapFactory.decodeStream(inputStream)
                    extracted(bitmap)
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                }
            }
        } else if (requestCode == REQ_CODE_CAMERA) {
            val img = intent!!.extras!!["data"] as Bitmap?
            extracted(img)
        }
    }

    private fun extracted(img: Bitmap?) {
        val text = Utils.dmtxDecode2(img)
        val textView = findViewById<TextView>(R.id.textView)
        textView.text = text
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }

    companion object {
        private const val REQ_CODE_PICK_IMAGE = 0
        private const val REQ_CODE_CAMERA = 1
        private const val RC_PERMISSION = 10
    }

    fun displayQR(qr: String) {
        val name = qr.toByteArray().sum().toString(26)
        MaterialDialog(this, BottomSheet()).show {
            customView(R.layout.qr_scanned)
            val scannedBinding = QrScannedBinding.bind(this.getCustomView())
            scannedBinding.watchName.text="Titan $name"
        }
    }


}